src/item/enchantment/AvailableEnchantmentRegistry.php

```diff
@@ -149,35 +149,42 @@ public function setSecondaryItemTags(Enchantment $enchantment, array $tags) : vo
 	}
 
 	/**
-	 * Returns enchantments that can be applied to the specified item.
-	 *
-	 * @param bool $onlyPrimaryTags Whether to check only primary item tags for each enchantment.
-	 *                              This should only be used to select available enchantments for an enchanting table.
+	 * Returns enchantments that can be applied to the specified item in an enchanting table.
 	 *
 	 * @return Enchantment[]
 	 */
-	public function getAvailableEnchantments(Item $item, bool $onlyPrimaryTags = false) : array{
+	public function getAvailableEnchantingTableEnchantments(Item $item) : array{
 		if(count($item->getEnchantmentTags()) === 0){
 			return [];
 		}
 
 		return array_filter(
 			$this->enchantments,
-			fn(Enchantment $enchantment) => $this->isAvailableForItem($enchantment, $item, $onlyPrimaryTags)
+			fn(Enchantment $e) => TagRegistry::getInstance()->isTagArraySubset($this->getPrimaryItemTags($e), $item->getEnchantmentTags())
 		);
 	}
 
 	/**
-	 * Returns whether the specified enchantment can be applied to the particular item.
+	 * Returns enchantments that can be applied to the specified item.
 	 *
-	 * @param bool $onlyPrimaryTags Whether to check only primary item tags for the enchantment.
-	 *                              This should only be used to select available enchantments for an enchanting table.
+	 * @return Enchantment[]
 	 */
-	public function isAvailableForItem(Enchantment $enchantment, Item $item, bool $onlyPrimaryTags = false) : bool{
-		$enchantmentTags = $this->getPrimaryItemTags($enchantment);
-		if(!$onlyPrimaryTags){
-			$enchantmentTags = array_merge($enchantmentTags, $this->getSecondaryItemTags($enchantment));
+	public function getAvailableEnchantments(Item $item) : array{
+		if(count($item->getEnchantmentTags()) === 0){
+			return [];
 		}
+
+		return array_filter(
+			$this->enchantments,
+			fn(Enchantment $enchantment) => $this->isAvailableForItem($enchantment, $item)
+		);
+	}
+
+	/**
+	 * Returns whether the specified enchantment can be applied to the particular item.
+	 */
+	public function isAvailableForItem(Enchantment $enchantment, Item $item) : bool{
+		$enchantmentTags = array_merge($this->getPrimaryItemTags($enchantment), $this->getSecondaryItemTags($enchantment));
 		return TagRegistry::getInstance()->isTagArraySubset($enchantmentTags, $item->getEnchantmentTags());
```

Comment:
> don't ask me why MASK is in both tag arrays, it just happens to be the case

I don't think we need this on the PM side, right? Anvils can apply any enchantment.

For the grouping problem, I would suggest resolving the group tag into all of its leaf tags, and then doing the intersection of the leaf nodes.

Reply:
> I don't think we need this on the PM side, right? Anvils can apply any enchantment.

Anvils can only apply enchantments that have primary and secondary tags covering the item tags, and are compatible with each other. But what does this have to do with my statement about the duplicate MASK tag? I meant that it could be the case that the plugin sets the enchantment tags in such a way that the primary and secondary tags individually cannot cover the item tags, my statement about MASK may have confused you. Another example, the primary enchantment tags are `[PICKAXE, AXE, FLINT_AND_STEEL]` and the secondary tags are `[SHOVEL, HOE, SHEARS]` and the item has `[DIG_TOOLS]`. DIG_TOOLS groups PICKAXE, AXE, SHOVEL, HOE, so enchantment can be applied to this item, since all item tags are among the enchantment tags. However, if we do the `isTagArraySubset([[PICKAXE, AXE, FLINT_AND_STEEL]], [DIG_TOOLS]) || isTagArraySubset([[SHOVEL, HOE, SHEARS]], [DIG_TOOLS]])` check, the result will be false. Why? Because we are checking primary and secondary tags separately (as you suggest), and neither array is a subset of the other. So we need to merge primary and secondary tags together first, and then call `isTagArraySubset([PICKAXE, AXE, FLINT_AND_STEEL, SHOVEL, HOE, SHEARS], [DIG_TOOLS])`, and then the result will be true, as it should be. Is that clear?

> For the grouping problem, I would suggest resolving the group tag into all of its leaf tags, and then doing the intersection of the leaf nodes.

This is exactly how I have implemented the grouping system already)